### PR TITLE
Introduce system.events

### PR DESCRIPTION
**Pull request by bhalevy**
Add a node-local system.events table to hold high-level events (typically administrative)
for the long term.

Events types include, for example:
- system start/stop
- keyspace and table create and drop
- major compaction
- repair

This series introduces a rudimentary implementation of such a table
and makes use of it for the above events.

Unit tests were added to test/cql-pytest for the above cases.

TODO: The schema for system.events in the series uses a text column to keep the event parameters, and each event passes a dictionary that is formatted into a string to keep them.

It would be better to keep the event params in a map collection so they will be stored in a more structured way and be searched more easily.

### CODE REVIEW

**Comment by xemul** [comment]
Cc @bentsi (AFAIR QA team nowadays parse system logs for certain events, likely they can benefit from that table)

**Comment by bhalevy** [comment]
@vladzcloudius please look into this from the field point of view

**Comment by slivne** [comment]
there are some redundant tables in the system

there is a compaction_history table 
there is a repair_history table 
there is a scylla_table_schema_history

So I think we need to better define what we are trying to collect - overall it may be ok to record items twice - or we may decide to change this to a virtual table that extract information from multiple tables.

**Comment by vladzcloudius** [comment]
@bhalevy Could you, please, clarify the context of this change?
It is nice and all, but we do have all this info in syslog today. Was there a feature request to implement such a table?

**Comment by bhalevy** [comment]
> @bhalevy Could you, please, clarify the context of this change? It is nice and all, but we do have all this info in syslog today. Was there a feature request to implement such a table?

syslog is not a reliable place to keep information.
The context is the sstable corruption issue we had in the past where partition got reordered when repair ran on nodes with different shard count.  When faced with a failure to write sstable in version 4.5 where we raised the validation bar we had no idea what versions ran on the cluster in the past, with how many cpus, when repair ran, etc.  This kind of information needs to be kept ling term for years, and some of it l, like version changes (upgrades, rollbacks) probably indefinitely.

It's a kind of an insurance policy, so we should learn from the past what could be useful to diagnose future problems.

**Comment by denesb** [comment]
> +1 We should avoid having the same info in multiple places. So, we should either kill old ones (but be careful - this may break some tools that use it) or drop this patch and make sure we properly update the existing ones.

If the new table has all the information the old ones had: we can replace the old ones with virtual tables and generate their content from the new one.

**Comment by bhalevy** [comment]
In v2 (366555a254)
- copy referenced input strings in `database::drop_column_family` since they were released when yielding for `start_system_event` - causing `database_test.database_drop_column_family_clears_querier_cache` to hang.

**Comment by bhalevy** [comment]
> > +1 We should avoid having the same info in multiple places. So, we should either kill old ones (but be careful - this may break some tools that use it) or drop this patch and make sure we properly update the existing ones.
> 
> If the new table has all the information the old ones had: we can replace the old ones with virtual tables and generate their content from the new one.

I the current version of this series it doesn't.

That said, we can probably move all the detailed information from the old table(s) (repair and compaction history for example), to system.events, with a shorter ttl for the high-resolution events (like per-range repair status or regular compaction jobs) and,
as you said, mimic the old tables using virtual tables.

**Comment by bhalevy** [comment]
@raphaelsc for a mixup of short and long ttls it could help to have variable-sized time windows rather than linear such that the window size would increase exponentially, e.g. for a basic window size of 1 day (as per the configuration), you'd keep up to 10 x 1-day windows, then 10 x 10-days windows, then 10 x 100-days windows (this covers about 3 years), then up to 10 x 1000 days windows for eternity

**Comment by bhalevy** [comment]
In v3 (7a4a458b15):
- cleaned up db::end_system_event + doc
- took start error handling in main out of the catch block so not to .get() in the catch block (anticipating we'll coroutinize main at some point)
  - used the auto start_event variable in main so it could be ended by both the good and bad paths.

File: repair/repair.cc [hunk 1]
```diff
@@ -1092,9 +1093,17 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
...
-        return id.id;
+        co_return id.id;
     }
 
+    db::event_params params = {
+        {"keyspace_name", keyspace},
+        {"options", format("{}", options_map)},
+        {"id", format("{}", id.id)},
+        {"uuid", format("{}", id.uuid)},
+    };
```
Comment (asias) [inline repair/repair.cc:1104]: Items I think we'd better add:

reason - reason for the repair (can be plain repair or node operations)
ranges - ranges requested to repair
tables - tables requested to repair

Currently, this patch ignores repair for repair based node operations (repair_service::do_sync_data_using_repair).  If we move the event into 
repair_tracker::run(), we can cover both repair and node ops automatically.

We do not have "options" for rbno repair. But repair options passed by user is not very useful,  perhaps we can ignore it even for regular repair.

File: db/system_events.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    if (!db::qctx) {
+        co_return system_event{};
+    }
+
+    auto using_ttl = ttl != NO_TTL ? format(" USING TTL {}", gc_clock::as_int32(ttl)) : "";
+    const sstring req = format("INSERT INTO {}.{} (category, started_at, type, params) VALUES (?, ?, ?, ?){}",
+            system_keyspace::NAME, system_keyspace::EVENTS, using_ttl);
+
+    // FIXME: for now, format all params into a string.
+    // Column should be turned into a map
```
Comment (avikivity) [inline db/system_events.cc:82]: These is no "for now" with tables.

File: db/system_keyspace.cc [hunk 3]
```diff
@@ -954,6 +955,34 @@ schema_ptr system_keyspace::discovery() {
     return schema;
 }
 
+schema_ptr system_keyspace::events() {
+    static thread_local auto schema = [] {
+        auto id = generate_legacy_id(NAME, EVENTS);
+        auto b = schema_builder(NAME, EVENTS, std::make_optional(id))
+            .with_version(generate_schema_version(id))
+            .with_column("category", utf8_type, column_kind::partition_key)
```
Comment (avikivity) [inline db/system_keyspace.cc:963]: It's impossible to get a log of events with this, because it's sorted the wrong way.

**Comment by avikivity** [comment]
I'm skeptical about the whole thing. Will we end up replicating every log message into this table? If not, we'll always prefer the log.

The log is always going to be easier to use and more reliable when the system is down.

File: repair/repair.cc (hunk 1 above)
Comment (vladzcloudius) [reply repair/repair.cc:1104]: > But repair options passed by user is not very useful, perhaps we can ignore it even for regular repair.

Why is not useful, @asias ? What "options" are we talking about here? Does include DCs whitelisting, for example? I think we every single parameter given to a repair task is important - even if you think it's not important to you now your opinion may change in the future.

**Comment by vladzcloudius** [comment]
> > @bhalevy Could you, please, clarify the context of this change? It is nice and all, but we do have all this info in syslog today. Was there a feature request to implement such a table?
> 
> syslog is not a reliable place to keep information. 

And a `system` KS is? 
You loose a node - you loose the info.
You decommission a node - `system` table data is lost.

What am I missing here?


The context is the sstable corruption issue we had in the past where partition got reordered when repair ran on nodes with different shard count. When faced with a failure to write sstable in version 4.5 where we raised the validation bar we had no idea what versions ran on the cluster in the past, with how many cpus, when repair ran, etc. This kind of information needs to be kept ling term for years, and some of it l, like version changes (upgrades, rollbacks) probably indefinitely.
> 
> It's a kind of an insurance policy, so we should learn from the past what could be useful to diagnose future problems.

@bhalevy, please, see my comments above. 
Unless I'm missing something your "insurance" is not much better than what syslog provides (rsyslog + syslog can actually provide even a better one that this PR).

To make it really reliable you should consider something better than `system` KS. `system_distributed_everywhere` would be a much better option for example.
The rate of writes in this KS are not supposed to be high therefore a high replication should not be an issue.

File: repair/repair.cc (hunk 1 above)
Comment (asias) [reply repair/repair.cc:1104]: I am talking about the options passed by the api. It is more useful to record the options after it is parsed. For example user may pass ranges, startToken, endToken, primaryRange to specify the actual ranges to repair. The ranges can even be empty which means all the ranges of the node. After parsing the options, we know which ranges the repair will repair no matter how the user specify it. Similarly, we have dataCenters / hosts / ignore_nodes to control which nodes to repair. If we just record the address of the nodes repair actually talks with, it is more informative. People who reads the table does not need to parse the options again.

**Comment by bhalevy** [comment]
Closing based on the feedback so far

**Pull request closed without merging**